Support separate payloads for RemoteCallbacks

omus opened the pull request:
Spin off of https://github.com/JuliaLang/julia/pull/26387. Support having separate payloads for each callback in `RemoteCallbacks`.  Specifically this change was made to support using the credential callback included in LibGit2.jl and the progress bar proposed in https://github.com/JuliaLang/Pkg3.jl/pull/177.

There is a little work yet to do including:
- [x] Documenting the new interface to `RemoteCallbacks` where tuples of callbacks/payloads are passed in
- [x] Tests

omus committed:
Support separate payloads for RemoteCallbacks

Make it easier to support mixing user created callbacks and default
callbacks defined by LibGit2.

omus commented:
cc: @KristofferC

omus commented on stdlib/LibGit2/src/types.jl:
```diff
@@ -206,6 +206,27 @@ abstract type Payload end
     payload::Ptr{Cvoid}
 end
 
+@kwdef struct CallbackPayload
```
I think I'll rename this to `RemotePayloads`

vchuravy commented on stdlib/LibGit2/src/LibGit2.jl:
```diff
@@ -275,8 +275,9 @@ function fetch(repo::GitRepo; remote::AbstractString="origin",
     result = try
-        fo = FetchOptions(callbacks=RemoteCallbacks(credentials=credentials_cb(), payload=p))
-        fetch(rmt, refspecs, msg="from $(url(rmt))", options = fo)
+        callbacks = RemoteCallbacks(credentials=(credentials_cb(), pointer_from_objref(p)))
```
`pointer_from_objref` is fishy here since it allows `p` to be gc'd, 
that's why original code uses `Ref`.

KristofferC replied:
`GC.@preserve` it?

vtjnash replied:
Too brittle and error-prone. This code really just needs to avoid using `Ptr{Void}` and stick with `Ref`.

vtjnash commented on stdlib/LibGit2/src/types.jl:
```diff
@@ -206,6 +204,27 @@ abstract type Payload end
+    push_transfer_progress::Ptr{Cvoid}
+    push_update_reference::Ptr{Cvoid}
+    push_negotiation::Ptr{Cvoid}
+    transport::Ptr{Cvoid}
```
These are Julia handles. They should not be typed as opaque C handles. It looks like the input is a `Dict`, so `RemotePayloads` can also just be a `Dict{Symbol, Any}`.

omus replied:
Should I use `::Any` or `Ref{Any}`. We could do either.

omus commented:
I can't seem to reproduce the CI failures locally. I'll push part of my work to try and sort out what's going wrong.

KristofferC commented:
Any update here?

omus commented:
Sorry, I've been on a work trip this last week. I'm planning on making this my highest priority when I get back.

omus committed:
Deprecate LibGit2 keyword: payload

LibGit2 functions `clone`, `fetch`, and `push` all used a keyword named
`payload` which would only take credential information. The keyword was
renamed to `credentials` to be more accurate and the original keyword
is now deprecated.

omus committed:
Callbacks documentation

omus committed:
Add test for credential callback conflict

omus committed:
Add TransferProgress struct

omus committed:
Add online callback test

omus commented:
macOS failure on Travis looks to be a DNS issue. I'll have some tests shortly for this PR.

StefanKarpinski commented:
@KristofferC, do you want to review this or should we just merge?
